File: nodejs/examples/extraction/03_schema_generation_sdk.ts
#!/usr/bin/env npx ts-node
/**
 * Schema Generation with SDK
 *
 * This example shows how to automatically generate CSS extraction schemas
 * from HTML using LLM. The schema can then be reused for fast, no-cost extraction.
 *
 * Usage:
 *     npx ts-node 03_schema_generation_sdk.ts
 *
 * Requirements:
 *     npm install crawl4ai-cloud
 */

import { AsyncWebCrawler } from 'crawl4ai-cloud';

// Configuration
const API_KEY = 'YOUR_API_KEY'; // Replace with your API key

async function generateExtractionSchema(): Promise<void> {
  const crawler = new AsyncWebCrawler({ apiKey: API_KEY });

  try {
    // First, get the HTML content
    console.log('Fetching Hacker News HTML...');
    const result = await crawler.run('https://news.ycombinator.com', {
      strategy: 'http',
    });

    const html = result.html || '';
    console.log(`Got ${html.length} bytes of HTML`);

    // Generate schema using LLM
    console.log('\nGenerating CSS extraction schema...');
    const schemaResult = await crawler.generateSchema(html, {
      query: 'Extract all stories with their title, URL, points, and author',
    });

    if (schemaResult.error) {
      console.log(`Error: ${schemaResult.error}`);
      return;
    }

    console.log('\nGenerated Schema:');
    console.log(JSON.stringify(schemaResult.schema, null, 2));

    // Now use the generated schema for extraction
    console.log('\n\nTesting generated schema...');
    const extractResult = await crawler.run('https://news.ycombinator.com', {
      strategy: 'http',
      config: {
        extraction_strategy: {
          type: 'json_css',
          schema: schemaResult.schema,
        },
      },
    });

    if (extractResult.success && extractResult.extractedContent) {
      const stories = JSON.parse(extractResult.extractedContent);
      console.log(`\nExtracted ${stories.length} stories`);
      console.log('\nFirst 2 stories:');
      stories.slice(0, 2).forEach((story: any) => {
        console.log(`\n${JSON.stringify(story, null, 2)}`);
      });
    }
  } finally {
    await crawler.close();
  }
}

generateExtractionSchema().catch(console.error);

File: python/examples/extraction/03_schema_generation_sdk.py
#!/usr/bin/env python3
"""
Schema Generation with SDK

This example shows how to automatically generate CSS extraction schemas
from HTML using LLM. The schema can then be reused for fast, no-cost extraction.

Usage:
    python 03_schema_generation_sdk.py

Requirements:
    pip install crawl4ai-cloud
"""

import asyncio
import json
from crawl4ai_cloud import AsyncWebCrawler, CrawlerRunConfig

# Configuration
API_KEY = "YOUR_API_KEY"  # Replace with your API key


async def generate_extraction_schema():
    """Generate CSS schema for Hacker News stories."""
    async with AsyncWebCrawler(api_key=API_KEY) as crawler:
        # First, get the HTML content
        print("Fetching Hacker News HTML...")
        result = await crawler.run(
            url="https://news.ycombinator.com",
            strategy="http"
        )

        html = result.html
        print(f"Got {len(html)} bytes of HTML")

        # Generate schema using LLM
        print("\nGenerating CSS extraction schema...")
        schema_result = await crawler.generate_schema(
            html=html,
            query="Extract all stories with their title, URL, points, and author"
        )

        if schema_result.error:
            print(f"Error: {schema_result.error}")
            return

        print("\nGenerated Schema:")
        print(json.dumps(schema_result.schema, indent=2))

        # Now use the generated schema for extraction
        print("\n\nTesting generated schema...")
        config = CrawlerRunConfig(
            extraction_strategy={
                "type": "json_css",
                "schema": schema_result.schema
            }
        )

        extract_result = await crawler.run(
            url="https://news.ycombinator.com",
            strategy="http",
            config=config
        )

        if extract_result.success and extract_result.extracted_content:
            stories = json.loads(extract_result.extracted_content)
            print(f"\nExtracted {len(stories)} stories")
            print("\nFirst 2 stories:")
            for story in stories[:2]:
                print(f"\n{story}")


if __name__ == "__main__":
    asyncio.run(generate_extraction_schema())
